fish-rust/src/env/environment_impl.rs

```diff
@@ -0,0 +1,1225 @@
+        // automatically be exported. But if you have a global pathvar, a local variable with the same
+        // name will be a pathvar. This is historical.
+        let mut flags = VarFlags::default();
+        if let Some(existing) = self.find_variable(key) {
+            flags.pathvar = Some(existing.is_pathvar());
+            flags.parent_exports = existing.exports();
+        }
+        if query.has_export_unexport {
+            flags.exports = Some(query.exports);
+        }
+        if query.has_pathvar_unpathvar {
+            flags.pathvar = Some(query.pathvar);
+        }
+
+        let mut result = ModResult::new(EnvStackSetResult::ENV_OK);
+        if query.has_scope {
+            // The user requested a particular scope.
+            // If we don't have uvars, fall back to using globals.
+            if query.universal && !UVAR_SCOPE_IS_GLOBAL.load() {
+                self.set_universal(key, val, query);
+                result.uvar_modified = true;
+            } else if query.global || (query.universal && UVAR_SCOPE_IS_GLOBAL.load()) {
+                Self::set_in_node(&mut self.base.globals, key, val, flags);
+            } else if query.local {
+                assert!(
+                    !self.base.locals.ptr_eq(&self.base.globals),
+                    "Locals should not be globals"
+                );
+                Self::set_in_node(&mut self.base.locals, key, val, flags);
+            } else if query.function {
+                // "Function" scope is:
+                // Either the topmost local scope of the nearest function,
+                // or the top-level local scope if no function exists.
+                //
+                // This is distinct from the unspecified scope,
+                // which is the global scope if no function exists.
+                let mut node = self.base.locals.clone();
+                while node.next().is_some() {
+                    node = node.next().unwrap();
+                    // The first node that introduces a new scope is ours.
+                    // If this doesn't happen, we go on until we've reached the
+                    // topmost local scope.
+                    if node.borrow().new_scope {
+                        break;
+                    }
+                }
+                Self::set_in_node(&mut node, key, val, flags);
+            } else {
+                panic!("Unknown scope");
+            }
+        } else if let Some(mut node) = Self::find_in_chain(&self.base.locals, key) {
+            // Existing local variable.
+            Self::set_in_node(&mut node, key, val, flags);
+        } else if let Some(mut node) = Self::find_in_chain(&self.base.globals, key) {
+            // Existing global variable.
+            Self::set_in_node(&mut node, key, val, flags);
+            result.global_modified = true;
+        } else if !UVAR_SCOPE_IS_GLOBAL.load()
```

Comment:
the !UVAR_SCOPE_IS_GLOBAL is a new invention. It seems correct but it's weird that it isn't used in the remove() function below

Reply:
Thanks, I'm not sure why I added that; removed.